### ISSUE TITLE
add 2d sprite rendering functionality

### ISSUE THREAD

**Issue by jakildev** [jakildev/IrredenEngine#14]
for hana <3

**Comment by jakildev** [comment]
Its pretty much in the title. We already have this functionality but I would like to codify it in the render code. Make it a legit path with proper game components, sprite sheets for animations, etc. And then we should create a demo that showcases some sprite rendering and animation. So we should have a sprite sheet animation system and component

**Comment by jakildev** [comment]
## Plan (proposed)

**TL;DR.** Decomposes into 5 stackable child tasks. State on the
ground is effectively greenfield — `system_sprites_to_screen.hpp` is
a copy-paste stub of `system_framebuffer_to_screen.hpp`, and
`entity_voxel_sprite.hpp` is entirely commented out — so we're
building clean, not patching.

### Child tasks

| # | Task | Area | Model | Depends on |
|---|------|------|-------|------------|
| 1 | Sprite data model + `C_Sprite` / `C_SpriteSheet` components (design note + prefab headers, no render code) | `engine/prefabs/irreden/render/`, docs | `[opus]` | — |
| 2 | Sprite-sheet asset format + loader (PNG + sidecar metadata, named sub-animations) | `engine/asset/`, `engine/prefabs/irreden/render/` | `[sonnet]` | 1 |
| 3 | `SPRITES_TO_TRIXEL` render pass — GLSL + MSL in lockstep | `engine/prefabs/irreden/render/systems/`, shaders | `[opus]` | 2 |
| 4 | `C_SpriteAnimation` + UPDATE-phase animation system (loop / once / ping-pong) | `engine/prefabs/irreden/render/` | `[opus]` | 1 (soft: 3) |
| 5 | Lua bindings + `sprite_demo` creation with `--auto-screenshot` shot list | `engine/script/`, `creations/demos/sprite_demo/` | `[sonnet]` | 2, 3, 4 |

Stack order: **1 → 2 → 3 → 4 → 5**.

### How it runs on the fleet

Each child lands as its own `fleet:task` issue with `Part of #14` in
the body. Once all have `human:approved` → queue-manager →
`T-A…T-E` in `TASKS.md`. Then:

```bash
fleet-claim stack "T-A T-B T-C T-D T-E" opus-worker-1
```

claims the chain atomically, writes a molecule yaml at
`~/.fleet/molecules/opus-worker-1.yml` for crash recovery, and the
worker processes them sequentially — each task becomes its own PR
with the previous task's branch as `--base`. Reviewers see
independently reviewable PRs; parent PRs in the chain pick up their
approvals individually.

### Open questions for the human (child 1 decides, but flag strong opinions)

1. **Canvas binding** — can a sprite live on any canvas, or only the
   active one?
2. **Lighting interaction** — sprite receives AO/shadow/light-volume
   modulation like voxels do, or bypasses lighting entirely?
3. **Sub-animation metadata format** — json sidecar vs. extend
   `IRAsset`'s text format with a new `.irsprite` extension?
4. **Lua API namespace** — `IRPrefab::Sprite::` by default per the
   layering principle in #266; different preference?

### Out of scope (flag as follow-ups if anyone asks)

- Sprite batching / instancing
- Animation blending / crossfade
- Sprite physics / collision
- Dynamic atlas repacking

---

Plan file saved at `~/.fleet/plans/issue-14.md` on the fleet host.
Ready to fanout into 5 child `fleet:task` issues on approval.

**Comment by jakildev** [comment]
## Plan revision — correcting the pipeline slot

The prior comment placed sprites as trixel content (`SPRITES_TO_TRIXEL`). That was wrong — sprites are **screen-composite layer, not trixel content**. The `SPRITE_TO_SCREEN` enum slot is the right home; the existing `system_sprites_to_screen.hpp` stub (currently a copy-paste of `system_framebuffer_to_screen.hpp`) gets replaced.

### v1 architecture locked in

- **Pipeline slot:** separate `SPRITES_TO_SCREEN` system at the `FRAMEBUFFER_TO_SCREEN` pipeline stage. Runs after the main canvas's composite tick, before the GUI canvas's tick. Sprites skip voxel→trixel→framebuffer entirely.
- **Depth model (v1):** sprites use `C_Position3D`; depth = iso-projection of the anchor, computed CPU-side for the sort.
- **Anchor:** `C_Sprite.anchor_` in local [0,1] UV space, default `{0.5, 0.0}` (bottom-center), configurable.
- **Z-sort scope:** sprites sort against each other only. Canvas framebuffers keep their fixed registration order (main/background/gui). Sprites draw on top of `main`, below `gui`. Cross-layer sort is Part 2.
- **Drawing:** one instanced draw per frame against `QuadVAOArrays` with a new `SpritesToScreenProgram`. Per-instance buffer holds transform, UV rect, texture index, tint.
- **Non-iso seam (future):** documented in child 1, not implemented. Later `C_Depth` + `C_Position2D` replaces the iso-projection call; sort key stays `float`.
- **Lighting:** sprites bypass all lighting passes (AO/shadow/light-volume/fog). They composite after the trixel pipeline, so they're not candidates for per-voxel lighting.

### Child tasks (unchanged shape, child 3 re-scoped)

| # | Task | Area | Model | Depends on |
|---|------|------|-------|------------|
| 1 | Sprite data model + `C_Sprite` / `C_SpriteSheet` components + design note | `engine/prefabs/irreden/render/`, docs | `[opus]` | — |
| 2 | Sprite-sheet asset format + loader | `engine/asset/`, `engine/prefabs/irreden/render/` | `[sonnet]` | 1 |
| 3 | **`SPRITES_TO_SCREEN` pass** — separate system at `FRAMEBUFFER_TO_SCREEN` stage, instanced draw, CPU z-sort among sprites | `engine/prefabs/irreden/render/systems/`, shaders | `[opus]` | 2 |
| 4 | `C_SpriteAnimation` + UPDATE-phase animation system | `engine/prefabs/irreden/render/` | `[opus]` | 1 (soft: 3) |
| 5 | Lua bindings + `sprite_demo` creation | `engine/script/`, `creations/demos/sprite_demo/` | `[sonnet]` | 2, 3, 4 |

Stack order: **1 → 2 → 3 → 4 → 5**.

### How it runs

Each child lands as its own `fleet:task` issue. On `human:approved` → queue-manager → `T-A…T-E` in `TASKS.md`. Then:

```bash
fleet-claim stack "T-A T-B T-C T-D T-E" opus-worker-1
```

atomically claims the chain, writes a molecule yaml for crash recovery, and the worker produces 5 stacked PRs with chained `--base`.

### Updated plan file

Saved at `~/.fleet/plans/issue-14.md` on the fleet host. Revision history at the bottom tracks this correction.

Filing the 5 child `fleet:task` issues now — will link them in a follow-up comment.

**Comment by jakildev** [comment]
## Child issues filed

The 5 children are now open in the backlog with `fleet:task`:

1. #282 — sprite rendering: data model + `C_Sprite` / `C_SpriteSheet` components `[opus]`
2. #283 — sprite rendering: sheet asset format + loader `[sonnet]`
3. #284 — sprite rendering: `SPRITES_TO_SCREEN` pass with instanced draw + iso z-sort `[opus]`
4. #285 — sprite rendering: `C_SpriteAnimation` + animation advance system `[opus]`
5. #286 — sprite rendering: Lua bindings + `sprite_demo` creation `[sonnet]`

Stack order: **#282 → #283 → #284 → #285 → #286**. Each child's body has its own acceptance criteria, scope, and dependency edges.

### Next step for the human

Add `human:approved` to each child as you're ready to release them into the queue. The queue-manager ingests approved issues into `TASKS.md` as `T-NNN` entries; once all five are there, a worker can:

```bash
fleet-claim stack "T-A T-B T-C T-D T-E" opus-worker-1
```

to claim the chain atomically and process the stacked PRs.

Parent #14 stays open as the epic tracker until all 5 children merge.

**Comment by jakildev** [comment]
Closed retroactively during open-issue audit. Completed by PRs #417, #494, #495, #507, #527 (sprite components, sheet loader, instanced draw, animation, ir.sprite.* Lua bindings, sprite_demo creation). These PRs predate the canonical `Closes #N` PR-body template added in #1042, so the issue was left open.
